packages/graphiql/src/state/GraphiQLSessionProvider.tsx

```diff
@@ -90,120 +98,117 @@ export const SessionContext = React.createContext<
-  React.useContext(SessionContext);
+export const useSessionContext = () => React.useContext(SessionContext);
 
-export function getSessionReducer(schema: GraphQLSchema) {
-  return (state: SessionState, { type: actionType, payload }: ActionState) => {
-    switch (actionType) {
-      case actionTypes.editor_loaded: {
-        const { context, editor } = payload as EditorPayload;
-        state.editors[context as EditorContexts] = editor as CodeMirror.Editor;
-        return state;
-      }
-      case actionTypes.operation_changed: {
-        const { value } = payload as ChangeValuePayload;
-        state.operation.text = value as string;
-        return {
-          ...state,
-          ...getQueryFacts(schema as GraphQLSchema, value),
-        };
-      }
-      case actionTypes.variables_changed: {
-        const { value } = payload as ChangeValuePayload;
-        state.variables.text = value as string;
-        return state;
-      }
-      case actionTypes.operation_succeeded: {
-        const { result } = payload as SuccessPayload;
-        state.results.text = result as string;
-        state.results.formattedText = JSON.stringify(result, null, 2);
-        state.operationErrors = null;
-        return state;
-      }
-      case actionTypes.operation_errored: {
-        const { error } = payload as ErrorPayload;
-        state.operationErrors = [error.toString()];
-        return state;
-      }
-      default: {
-        return state;
-      }
+function reducer(state: SessionState, action: Action): SessionState {
+  switch (action.type) {
+    case actionTypes.editor_loaded: {
+      const { context, editor } = action.payload;
+      return {
+        ...state,
+        editors: {
+          ...state.editors,
+          [context]: editor,
+        },
+      };
     }
-  };
+    case actionTypes.operation_changed: {
+      return {
+        ...state,
+        operation: {
+          ...state.operation,
+          text: action.payload,
+        },
+        // TODO: do this in a useMemo where it's needed
+        // ...getQueryFacts(schema, value),
+      };
+    }
+    case actionTypes.variables_changed: {
+      return {
+        ...state,
+        variables: {
+          ...state.variables,
+          text: action.payload,
+        },
+      };
+    }
+    case actionTypes.operation_succeeded: {
+      return {
+        ...state,
+        operationErrors: null,
+        results: {
+          ...state.results,
+          text: action.payload.result,
+        },
+      };
+    }
+    case actionTypes.operation_errored: {
+      return {
+        ...state,
+        operationErrors: [String(action.payload)],
+      };
+    }
+    default: {
+      return state;
+    }
+  }
 }
 
 export type SessionProviderProps = {
   sessionId: number;
   fetcher?: Fetcher;
   session?: SessionState;
-  children: any;
+  children: React.ReactNode;
 };
 
 export function SessionProvider({
```

Comment:
for `SessionProvider`, wondering if we should (eventually) consume the three editor instances, as we will use them for handling change events that way with monaco. each editor has `editor.onDidChangeContent` or some such, so we can use those when we export the handlers for operation/variables change and such?

Reply:
absolutely! I'll tinker on consuming these today